doc/BluetoothLEStack.md

```diff
@@ -0,0 +1,171 @@
+Bluetooth Low Energy Design Document
+====================================
+
+## System call interface
+
+The system call interface is modeled after the HCI interface defined in the
+Bluetooth specification.
+
+### Device address
+
+The kernel assigns the device address. The process may read the device address
+using an `allow` system call.
+
+### Advertising
+
+For advertising, the sytsem call interface allows a process to configure an
+advertising payload, advertising event type, scan response payload, interval and
+tx power. Permissible advertising types include:
+
+  * Connectable undirected
+
+  * Connectable directed
+
+  * Non-connectable undirected
+
+  * Scannable undirected
+
+The driver is _not_ responsible for validating that the payload for these
+advertising types follows any particular specification. Advertising event types
+that require particular interactions at the link-layer with peer devices (e.g.
+scanning or establishing connections) are not permissible:
+
+  * Scan request
+
+  * Scan response
+
+  * Connect request
+
+Scan response are sent automatically if a scan response payload is configured.
+Scan rerequest and connection requests are handled by other parts of the sytsem
+call interface.
+
+To set up an advertisement:
+
+  1. Configure the advertisement payload, type, interval, tx power and,
+     optionally, scan response payload.
+
+     * Advertisement payload `allow`
+
+     * Advertisement type `command`
+
+     * If the advertising type is scannable, you SHOULD configure a scan
+       response payload using `allow`
+
+     * Advertisement interval `command`
+
+     * Advertisement tx power `command`
+
+  2. Start periodic advertising using a `command`
+
+Any changes to the configuration while periodic advertising is happening will
+take effect in a future advertising event. The kernel will use best effort to
+reconfigure advertising in as few events as possible.
+
+To stop advertising
+
+  1. Stop periodic advertising a `command`
+
+### Scanning
+
+### Connection-oriented communication
+
+## Hardware Interface Layer (HIL)
+
+The Bluetooth Low Energy Radio HIL defines a cross-platform interface for
+interacting with on-chip BLE radios (i.e. it does not necessarily work for
+radios on a dedicated IC connected over a bus).
+
+The goal of this interface is to expose low-level details of the radio that are
+common across platforms, except in cases where abstraction is needed for common
+cases to meet timing constraints.
+
+
+```rust
+pub trait BleRadio {
+    /// Sets the channel on which to transmit or recieve packets.
+    ///
+    /// Returns ReturnCode::EBUSY if the radio is currently transmiting or
+    /// recieving, otherwise ReturnCode::Success.
+    fn set_channel(&self, channel: RadioChannel) -> ReturnCode;
```

Comment:
Can you give an example how this intended to work? 

I would rather see the following sequence:
1. start_advertise(buf, channel).
2. continue_advertise(channel)

This would passing/copying a advertisement_buffer everytime we should receive or transmit advertisements when in practice we only need to change channel!

Reply:
Why not simply `advertise(buf)` which would perform advertisements on all three channels?

Reply:
For efficiency reasons, no reason to copy the buffer when we have `mutual exclusion` if we are switching channel from `37 -> 38``or `38 -> 39` but could probably solved internally in the method otherwise. But, currently it's like you proposed :)

Reply:
I think the sequence could be more verbose in that case :)

1. set_busy(true)
2. set_buffer(buf)
3. advertise_on(37)
4. advertise_on(38)
5. advertise_on(39)
6. set_busy(false)

Reply:
Sure, that is way cleaner